R/get-email.R

```diff
@@ -1,18 +1,76 @@
-## takes a Token2.0, extracts access token, retrieves user's email
-## does not (cannot, really) refresh said token
-## use it when you expect the token to be valid for external reasons
-## also, assumes the token was obtained with "email" scope
-## this will be true of all gargle-mediated tokens, by definition
-get_email <- function(token) {
+# Good overview re: learning about your Google user / token:
+# https://www.oauth.com/oauth2-servers/signing-in-with-google/verifying-the-user-info/
+#
+# It suggests perhaps we could recover, e.g., the email during token initiation?
+# But I'm not sure and it might require a change in httr, in any case.
+#
+# Therefore, I focus on the two post hoc methods for retrieving info based on
+# the token you've already got:
+#   * userinfo endpoint
+#   * tokeninfo endpoint
+# Both require an additional HTTP request, so are "not recommended for
+# production applications".
+#
+# Google's docs re: how to get user info:
+# https://developers.google.com/identity/protocols/OpenIDConnect#obtaininguserprofileinformation
+# "Add your access token to the authorization header and make an HTTPS GET
+# request to the userinfo endpoint, which you should retrieve from the Discovery
+# document using the key userinfo_endpoint."
+
+# Here's the OpenID Connect discovery document:
+# https://accounts.google.com/.well-known/openid-configuration
+#
+# Here's the URL for userinfo_endpoint, at the time of writing:
+# https://openidconnect.googleapis.com/v1/userinfo
+#
+# I had working code to access both of these endpoints before I read this and
+# there are a few differences between what I do and what they say to do. The
+# discrepancies are around which token to send (ID? access? etc.) and where/how
+# to send it. I conclude that there must be a few variations that work.
```

Comment:
So id tokens are a different way to prove your identity, as they're self-signed tokens that anyone can verify (and implement the openid connect "spec").

Some details are [in the docs](https://developers.google.com/identity/protocols/OpenIDConnect#exchangecode), but roughly speaking, this is my understanding:

* if I grant the `userinfo.email` scope, the service I'm talking to can directly talk to google to determine my identity

* if I create an id token, I can hand that to a service, and they can verify with google that this is a valid token

Both play similar roles; my impression is that the first is more useful for us here, since we're in the business of talking to one or more google APIs, and the second is more useful in a situation where Google is just one of many identity providers a server may want to use.

I don't have a huge amount of experience with id tokens, but one use case (that may be interesting for Shiny?) that's worth calling out: if you have both a frontend and a backend, and the user is going to authenticate to your frontend, it's nice to avoid them doing another auth step against your backend. One way to make this work is to have your frontend get an id token, which you then present to the backend with the request; the backend can verify the identity based on that. (However, this only solves a problem if the backend doesn't need additional scopes, or is willing to walk you through additional flows for them later.)

Reply:
Thanks for this. Glad you also agree that userinfo.email is the more useful route for us. I'll file away this knowledge about ID tokens and possible relevance to Shiny.